### ISSUE TITLE
FileManagerTest::VerifyFlushedFileContent  `deduceType` ERROR

### ISSUE THREAD

**Issue by kkli08** [kkli08/KV-Store#49]
```c++
TEST(FileManagerTest, VerifyFlushedFileContent) {
    FileManager fileManager(fs::path("test_db"));
    std::vector<KeyValue> kv_pairs = { KeyValue(1, 100), KeyValue(2, 200) };

    FlushSSTInfo info = fileManager.flushToDisk(kv_pairs);

    // Open the SST file and check its content
    std::ifstream file("test_db/" + info.fileName, std::ios::binary);
    ASSERT_TRUE(file.is_open());

    // Deserialize the header
    SSTHeader header = SSTHeader::deserialize(file);
    EXPECT_EQ(header.num_key_values, 2);

    // Deserialize and check the first KeyValue pair
    SerializedKeyValue skv1 = SerializedKeyValue::deserialize(file);
    EXPECT_EQ(std::get<int>(skv1.kv.getKey()), 1);
    EXPECT_EQ(std::get<int>(skv1.kv.getValue()), 100);

    // Deserialize and check the second KeyValue pair
    SerializedKeyValue skv2 = SerializedKeyValue::deserialize(file);
    EXPECT_EQ(std::get<int>(skv2.kv.getKey()), 2);
    EXPECT_EQ(std::get<int>(skv2.kv.getValue()), 200);

    file.close();

    // Clean up: Delete the test directory
    fs::remove_all("test_db");
}
```

**Comment by kkli08** [comment]
### The Problem

The core of the problem was that the `deduceType` function was expecting primitive types like `int`, `double`, `std::string`, etc., but it was receiving a `std::variant` instead. `std::variant` is a container that can hold multiple types, but it's not a primitive type itself. When `deduceType` tried to deduce the type of the `std::variant`, it failed because `std::variant` wasn't one of the types that `deduceType` could handle. This is what caused the "Unsupported type" exception.

The key issue was that, when the constructor of `KeyValue` was invoked during deserialization, it was passing a `std::variant` object to `deduceType`. But `deduceType` didn't know how to handle a `std::variant`, because it was designed to handle basic types like `int`, `double`, etc.

### The Fix

The fix involved using `std::visit` to unpack the `std::variant` and pass the underlying type (such as `int`, `double`, `std::string`, etc.) to `deduceType`. `std::visit` is a C++17 feature that allows you to access the actual type stored inside a `std::variant`.

### How It Works

1. **`std::variant`:** 
   - `std::variant` can hold multiple types (in your case: `int`, `long long`, `double`, `char`, `std::string`). However, you can't directly pass a `std::variant` to a function that expects a specific type like `int` or `double`, because `std::variant` can hold different types at runtime.

2. **`std::visit`:**
   - `std::visit` allows you to access the actual value stored in a `std::variant` and apply a function (or lambda) to that value. This way, you can apply logic based on the specific type inside the `std::variant`.
   
   - In your case, we use `std::visit` to pass the actual type inside the `std::variant` (like `int`, `std::string`, etc.) to `deduceType`. This way, `deduceType` can correctly deduce and return the appropriate enum (`KeyValueType::INT`, `KeyValueType::STRING`, etc.).

### Key Points:

- **Original Issue:**
   - `deduceType` was being called on a `std::variant`, which is a container for multiple types. Since `std::variant` is not a recognized primitive type in `deduceType`, the function threw an "Unsupported type" exception.
   
- **Solution:**
   - We used `std::visit` to unpack the `std::variant` and pass the actual stored type (like `int`, `std::string`, etc.) to `deduceType`. This allowed `deduceType` to handle the actual type correctly and return the appropriate `KeyValueType` enum.

### What Happens Now

1. When you call the `KeyValue` constructor:
   ```cpp
   skv.kv = KeyValue(key, value);
   ```
   The `key` and `value` are `std::variant`s holding the actual data types (like `int`, `std::string`, etc.).

2. The constructor now uses `std::visit`:
   ```cpp
   keyType = std::visit([&](auto&& arg) {
       return deduceType(arg);  // Deduce the actual type of the key
   }, key);
   ```
   This unpacks the `std::variant` and passes the actual type (`arg`, which could be an `int`, `std::string`, etc.) to `deduceType`.

3. The `deduceType` function can now correctly identify the type (because it’s receiving the actual type and not the `std::variant`) and returns the appropriate `KeyValueType`.

4. The same process is applied for both the `key` and the `value`.

### Final Takeaway

In summary, the issue was due to passing the `std::variant` (which is a container for multiple types) directly to `deduceType`. Using `std::visit` allows us to extract the actual type from the `std::variant` and pass that to `deduceType`, which solved the problem.

**Comment by kkli08** [comment]
## Previous KeyValue Constructor Template
```c++
template<typename K, typename V>
KeyValue::KeyValue(K k, V v) : 
                    key(convertToVariant(k)), 
                    value(convertToVariant(v)), 
                    keyType(deduceType(k)), 
                    valueType(deduceType(v)) 
                    {}
```

## Updated KeyValue Constructor Template
```c++
template<typename K, typename V>
KeyValue::KeyValue(K k, V v)
    : key(convertToVariant(k)), value(convertToVariant(v)) {
    keyType = std::visit([&](auto&& arg) {
        return deduceType(arg);  // Deduce the actual type of the key
    }, key);
    valueType = std::visit([&](auto&& arg) {
        return deduceType(arg);  // Deduce the actual type of the value
    }, value);
}
```
